node-js-basics-101/node_modules/asynquence-contrib/contrib.src.js

```diff
@@ -0,0 +1,2360 @@
+									if (err) {
+										ret = it["throw"](err);
+										err = null;
+									}
+									// resume normally
+									else {
+											ret = it.next(msg);
+										}
+								}
+								// resumption failed, so bail
+								catch (e) {
+									done = true;
+									err = e;
+									msg = null;
+									unblock();
+									return;
+								}
+
+								// keep track of the result of the resumption
+								done = ret.done;
+								ret = ret.value;
+								type = typeof ret;
+
+								// if this goroutine is complete, unblock the
+								// overall goroutine handling
+								if (done) {
+									unblock();
+								}
+
+								// received a thenable/promise back?
+								if (isPromise(ret)) {
+									ret = ASQ().promise(ret);
+								}
+
+								// wait for the value?
+								if (ASQ.isSequence(ret)) {
+									ret.val(function $$val() {
+										ret = null;
+										msg = arguments.length > 1 ? ASQ.messages.apply(ø, arguments) : arguments[0];
+										next();
+									}).or(function $$or() {
+										ret = null;
+										msg = arguments.length > 1 ? ASQ.messages.apply(ø, arguments) : arguments[0];
+										if (msg instanceof Error) {
+											err = msg;
+											msg = null;
+										}
+										next();
+									});
+								}
+								// immediate value, prepare it to go right back in
+								else {
+										msg = ret;
+										ret = null;
+										next();
+									}
+							}
+						})();
+
+						// keep this goroutine alive until completion
+
+					case 7:
+						if (done) {
+							context$3$0.next = 17;
+							break;
+						}
+
+						context$3$0.next = 10;
+						return token;
+
+					case 10:
+						if (!(!done && !token.block && token.unblock_count === 0)) {
+							context$3$0.next = 14;
+							break;
+						}
+
+						context$3$0.next = 13;
+						return token.block = ASQ.iterable();
+
+					case 13:
+
+						token.block = false;
+
+					case 14:
+
+						if (token.unblock_count > 0) token.unblock_count--;
+						context$3$0.next = 7;
+						break;
+
+					case 17:
+
+						// this goroutine is done now
+						token.go_count--;
+
+						// all goroutines done?
+						if (token.go_count === 0) {
+							// any lingering blocking need to be cleaned up?
+							unblock();
+
+							// capture any untaken messages
+							msg = ASQ.messages.apply(ø, token.messages);
+
+							// need to implicitly force-close channel?
+							if (token.channel && !token.channel.closed) {
+								token.channel.closed = true;
+								token.channel.put_queue.length = token.channel.take_queue.length = 0;
+								token.channel.close = token.channel.go = token.channel.messages = null;
+							}
+							token.channel = null;
+						}
+
+						// make sure leftover error or message are
+						// passed along
+
+						if (!err) {
+							context$3$0.next = 23;
+							break;
+						}
+
+						throw err;
+
+					case 23:
+						if (!(token.go_count === 0)) {
```

Comment:
__[jshint]__
* Confusing use of '!'.

<!-- Comment by [SideCI](https://sideci.com) -->